Added the sympyfy function alongside its necessary dependencies

Collins-kariuk opened the pull request:
The sympyfy function outputs a reduced analytical or numerical Hamiltonian in momentum space utilizing the Fourier series. The input of this function is a TRIQS tight-binding model that one can get from converting their model from either a Wannier90 model or PythTB model using TRIQS's handy function that enables said conversion.

This function has been tested on 2 compounds with different symmetries. These compounds include a Strontium compound (SrCaCu2O4) and a Palladium compound (PdCoO2). This function mainly follows from the equation in the attached screenshot. This function will help anyone, especially experimentalists, better understand their results in the tight-binding model landscape by visualizing a reduced analytical expression (and a numerical expression should they wish).
![Tight-Binding Hamiltonian in Momentum-Space from Real Space using the Fourier Series](https://github.com/TRIQS/triqs/assets/74094734/4ec879c5-9d60-4d1b-a9bb-c89f7a4dd7bf)

Collins-kariuk committed:
Added the sympyfy function alongside its necessary dependencies

Added the sympyfy function into the utils.py file. In a nutshell, this function returns the analytical form of the momentum space hamiltonian of the tight-binding model from a tight-binding lattice object, by utlizing Fourier series.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+    
+    # obtaining unit vectors
+    # reduce floating point precision to 3 decimal places
+    a1 = np.around(TB_lat_obj_units_transpose[0], decimals = 3)
```
does it make sense to let the user define the precision at this point?

Collins-kariuk replied:
The default according to Wannier is 10^-6 but the user can change it, there will be an optional parameter to change the precision.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+    
+    Returns
+    -------
+    Hk_numerical: NumPy Array
```
This looks as if the function gives back to values, but it only gives back one value. Just list in Return one object and explain in the text that the object given back depends on the `analytical` flag.

Collins-kariuk replied:
Fixed.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+from sympy import *
+import warnings
+
+def sympyfy(tb_lat_obj, analytical = True):
```
I am not sure what is the meaning of the numerical form? Just the Fourier transform of H(R)? On what kind of mesh?

Collins-kariuk replied:
It is the Fourier transform of H(R) which only depends on **k**. There is no mesh because **k** is not specified. 
In some sense, it is also analytical because it depends on trigonometric functions but there are no lattice parameters or displacement vectors.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+from pythtb import *
+# from triqs.lattice.tight_binding import TBLattice # TRIQS needs to be installed for this dependency to work
+import sympy as sp
+from sympy import *
```
It would be good to include import module statements needed for the function in the function itself in the beginning, and more general things like itertools, and warnings at the top of the `uitls.py` file.

Collins-kariuk replied:
Fixed.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+    # converting the numerical Hamiltonian to a NumPy array from a SymPy matrix
+    Hk_numerical = np.array(Hk_numerical)
+
+    def has_complex_exponential(matrix):
```
maybe clarify that the functions are only for sympy by adding sympy to the function name. Consider making the functions iternal only by adding `_` in front.

Collins-kariuk replied:
Fixed.

the-hampel commented on python/triqs/lattice/utils.py:
```diff
@@ -246,3 +246,161 @@ def TB_from_pythTB(ptb):
+    for key, hopping in TB_lat_obj_hops.items():
+        rx, ry, rz = key
+        # reduce floating point precision of hopping parameters to 3 decimal places
+        hopping = np.around(hopping, decimals = 3)
```
Does it make sense to let the user specify the precision?

the-hampel reviewed (changes requested):
Hi @Collins-kariuk ,
thank you for the PR. Please consider my comments below. Also could you please create the PR against the triqs/unstable branch instead. Thank you.

Collins-kariuk committed:
Included some of Alex Hampel's recommendations. See the description:

- Updated where I called the import functions. I placed the sympy importations inside the sympfy function.
- updated the internal functions' names.
- updated the dosctring to include only 1 return value

Collins-kariuk committed:
included last recommendations from Alex Hampel

Collins-kariuk commented:
Olivier, Sophie, and I thought it best if I create a new branch, make the necessary changes to utils.py within that branch, and make a pull request from my branch to the upstream unstable branch. So I am going to close this pull request.
